src/compute/src/render/reduce.rs

```diff
@@ -105,29 +101,55 @@ where
-        ReducePlan::Basic(expr) => build_basic(collection, expr).into(),
+        ReducePlan::Hierarchical(HierarchicalPlan::Monotonic(expr)) => {
+            let (output, errs) = build_monotonic(debug_name, collection, expr);
+            err_collection = err_collection.concat(&errs);
+            output.into()
+        }
+        ReducePlan::Hierarchical(HierarchicalPlan::Bucketed(expr)) => {
+            let (output, errs) = build_bucketed(debug_name, collection, expr);
+            if let Some(e) = errs {
+                err_collection = err_collection.concat(&e);
+            }
+            output.into()
+        }
+        ReducePlan::Basic(BasicPlan::Single(index, aggr)) => {
+            build_basic_aggregate(debug_name, collection, index, &aggr).into()
+        }
+        ReducePlan::Basic(BasicPlan::Multiple(aggrs)) => {
+            build_basic_aggregates(debug_name, collection, aggrs).into()
+        }
         // Otherwise, we need to render something different for each type of
         // reduction, and then stitch them together.
         ReducePlan::Collation(expr) => {
             // First, we need to render our constituent aggregations.
             let mut to_collate = vec![];
 
-            if let Some(hierarchical) = expr.hierarchical {
-                to_collate.push((
-                    ReductionType::Hierarchical,
-                    build_hierarchical(collection.clone(), hierarchical),
-                ));
-            }
-            if let Some(accumulable) = expr.accumulable {
-                let (arranged_output, errs) =
-                    build_accumulable(debug_name, collection.clone(), accumulable);
-                err_collection = err_collection.concat(&errs);
-                to_collate.push((ReductionType::Accumulable, arranged_output));
-            }
-            if let Some(basic) = expr.basic {
-                to_collate.push((ReductionType::Basic, build_basic(collection.clone(), basic)));
+            for plan in [
+                expr.hierarchical.map(ReducePlan::Hierarchical),
+                expr.accumulable.map(ReducePlan::Accumulable),
+                expr.basic.map(ReducePlan::Basic),
+            ]
+            .into_iter()
+            {
+                let Some(plan) = plan else { continue };
+                let r#type = ReductionType::try_from(&plan)
+                    .expect("only representable reduction types were used above");
+
+                let (oks, errs) = render_reduce_plan_inner(
+                    debug_name,
+                    plan,
+                    collection.clone(),
+                    err_collection,
+                    key_arity,
+                );
+                let ArrangementOrCollection::Arrangement(arrangement) = oks else {
+                    panic!("interior reduce plans must produce an arrangement")
+                };
+                err_collection = errs;
```

Comment:
It turns out `DistinctNegated` is the only case that doesn't return an arrangement.  So I think this can actually be much cleaner and `ArrangementOrCollection` can go away entirely.

Reply:
I think the problem here is that we would like to avoid forcing an arrangement on the output of `DistinctNegated`, so we need the construction with `ArrangementOrCollection`, no? If we force that arrangement to be built, we'd be potentially keeping around a lot more state than we need to for some plans.

If you think the construction with the trait could work, that might help simplify the code somewhat without touching the distinct rendering variants.

Reply:
Of course, this change isn't necessary, so I can drop it entirely if we can't find consensus on it.

Reply:
+1 for removing `ArrangementOrCollection`, which I introduced as a means to de-duplicate code. It's clearly not needed anymore after this change.

Reply:
Also fine from my side: I like that we get rid of `ArrangementOrCollection` and at the same time do not arrange after `DistinctNegated`. Also it's much clearer to see now that `DistinctNegated` treatment is really special-cased.